README.md
# gawk-aregex

Gawk extension for approximate regex (fuzzy) matching, using the TRE
regex library from the [TRE](https://laurikari.net/tre/) library (also [here](https://github.com/laurikari/tre)).

Provides an `amatch()` function, roughly equivalent to the built-in
`match()` function in gawk. For documentation of this function and
example usage, please see the [man](doc/aregex.md) page.

As of 2019-01-18, this gawk extension is now incorporated into the
combined [Gawkextlib](http://gawkextlib.sourceforge.net/) project, and
appears [here](https://sourceforge.net/projects/gawkextlib/files/).

## Installation

 * (Install gawk, version 4.2+)
 * Install [gawkextlib](http://gawkextlib.sourceforge.net/) ([AUR](https://aur.archlinux.org/packages/gawkextlib/), [Fedora](https://apps.fedoraproject.org/packages/gawkextlib-devel)). Tested with version 1.0.4.
 * Install [TRE](https://laurikari.net/tre/) ([Arch](https://www.archlinux.org/packages/community/x86_64/tre/), [Fedora](https://apps.fedoraproject.org/packages/tre-devel)). Tested with version 0.8.0.
 * Make sure the libraries (`libgawkextlib.so` and `libtre.so`) and
   header files (`gawkapi.h` and `tre/tre.h`) can be found by the
   compiler. Add `-Ldir` and `Idir` arguments to `gcc` if needed.
 * Compile with `make`
 * Test with `make check`
 * Set install location `PREFIX` in `Makefile`. Install with `make install`

### Alternative installation

The file `aregex.c` can be easily incorporated into the standard
gawkextlib build chain:

      git clone git://git.code.sf.net/p/gawkextlib/code gawkextlib-code
      cd gawkextlib-code
      ./make_extension_directory.sh -g /.../local/bin/ -l /.../local/lib/ \
        -I aregex "Name" "email"
      cd aregex
      cp -f .../aregex.c .
      sed -i '7 i \#include "common.h"' aregex.c 
      ./configure # --prefix=/.../local/ 
      sed -i 's/-lgawkextlib/-lgawkextlib -ltre/g' Makefile
      make
      make install

The [Makefile](sf_build/Makefile) in `sf_build` will do this.

### Windows users

There is an up-to-date (version 5+) `gawk` for Windows at the
[ezwinports](https://sourceforge.net/projects/ezwinports/) project. I
looked briefly into cross-compiling `aregex.c` with
`i686-w64-mingw32`, but this would require also cross-compiling TRE
and gawkextlib, which is beyond me (given time available). Windows
users will need to locate a Linux or Windows machine.


## A note on bytes and characters

While the `amatch()` function is roughly equivalent to the gawk
`match()` function, I chose not to return `[i,"start"]` position and
`[i,"length"]` in the returned substring array (e.g., see [here](https://github.com/camwebb/gawk-aregex/blob/115e400dbe6446d3138c7da2fba6d461026cad8d/aregex.c)),
but to return just the literal substring for each parenthetical
match. Gawk is multibyte aware, and `match()` works in terms of
characters, not bytes, but TRE seems not to be character-based. Using
the `wchar_t` versions of `tre_regcomp()` and `tre_regaexec()` does
not help if the input is a mix of single and multi-byte characters.

A simple routine must be used on the output array (`out`), if
positions and lengths of the substrings are needed:

      print "i", "substring", "posn", "length"
      p = 1
      for (i = 1; i < length(out); i++) {
        idx = index(substr(str, p), out[i])
        len = length(out[i])
        print i, out[i], idx+p-1, len
        p = p + idx + len
      }

## Thanks to...

 * Ville Laurikari (@laurikari) for TRE
 * Arnold Robbins for maintaining Gawk
 * `gawkextlib` developers, and the developers of other extensions for
   their examples
 * Benjamin Eckel (@bhelx) for [this gist](https://gist.github.com/bhelx/1498622/e442385814e64625cc286ad3fa4ef9c0517ce785).
 * user sashoalm on StackOverflow for [this answer](https://stackoverflow.com/a/35200252/563709).
 * User Stefan on StackOverflow for [this answer](https://stackoverflow.com/a/17764716/563709).

----

Cam Webb <cw@camwebb.info>, 2018-11-24


ChangeLog
2019-05-13  Cam Webb <cw@camwebb.info>
	
	* aregex.c (main): fixed major memory leak. Needed to add tre_regfree()
	Found with: valgrind --tool=memcheck --leak-check=full gawk -f
	test/aregex.awk

2019-01-14  Cam Webb  <cw@camwebb.info>

	* NEWS: First version committed to Sourceforge.



test/aregex.awk
@load "./aregex.so"

BEGIN {
  str = "abcdễfbc"
  print "String    : " str
  re = "^a(bc)d(ễ)(f)$"
  print "RE        : " re

  # Set some costs
  cost["max_cost"] = 6
  print "max_cost  : " cost["max_cost"]
  cost["cost_ins"] = 2
  cost["cost_del"] = 3

  print "Match sr  : " amatch(str, re)
  print "Match src : " amatch(str, re, 6.1)
  print "Match srcb: " amatch(str, re, 6, b), b[1] 
  print "Match srC : " amatch(str, re, cost)
  print "Match srCb: " amatch(str, re, cost, out)

  print "Submatches: "
  print "   i  substr    pos  len"
  print "   -  ------    ---  ---"
  printf "  %2d  %-10s %2d %2d\n", 0, out[0], index(str, out[0]), length(out[0])
  p = 1
  for (i = 1; i < length(out); i++) {
    idx = index(substr(str, p), out[i])
    len = length(out[i])
    printf "  %2d  %-10s %2d %2d\n", i, out[i], idx+p-1, len
    p = p + idx + len
  }

  print "cost      : " cost["cost"]
  print "num_ins   : " cost["num_ins"]
  print "num_del   : " cost["num_del"]
  print "num_subst : " cost["num_subst"]
}


test_aregex.awk
@load "./aregex.so"

BEGIN {
  str    = "abcdefghi"
  cost   = 2

  re     = "^bcdefghij$"
  print "RE1: '" re "', string: '" str "', cost: " cost ", match dist: " \
    aregex(re, str, cost)
  # RE1: '^bcdefghij$', string: 'abcdefghi', cost: 2, match dist: 2
  
  re     = "^xcdefghij$"
  print "RE2: '" re "', string: '" str "', cost: " cost ", match dist: " \
    aregex(re, str, cost)
  # RE2: '^xcdefghij$', string: 'abcdefghi', cost: 2, match dist: -1

  # generate warning with gawk --lint :
  print "RE3: '" re "', string: '" str "', cost: " cost ", match dist: " \
    aregex(re, str, cost, "extra1", "extra2")
}


test/test_aregex.awk
@load "./aregex.so"

BEGIN {
  str = "abcdễfbc"
  print "String    : " str
  re = "^a(bc)d(ễ)(f)$"
  print "RE        : " re

  # Set some costs
  cost["max_cost"] = 6
  print "max_cost  : " cost["max_cost"]
  cost["cost_ins"] = 2
  cost["cost_del"] = 3

  print "Match(2)  : " amatch(str, re)
  print "Match(3)  : " amatch(str, re, cost)
  print "Match(4)  : " amatch(str, re, cost, out)

  print "Submatches: "
  print "   i  substr    pos  len"
  print "   -  ------    ---  ---"
  printf "  %2d  %-10s %2d %2d\n", 0, out[0], index(str, out[0]), length(out[0])
  p = 1
  for (i = 1; i < length(out); i++) {
    idx = index(substr(str, p), out[i])
    len = length(out[i])
    printf "  %2d  %-10s %2d %2d\n", i, out[i], idx+p-1, len
    p = p + idx + len
  }

  print "cost      : " cost["cost"]
  print "num_ins   : " cost["num_ins"]
  print "num_del   : " cost["num_del"]
  print "num_subst : " cost["num_subst"]
}


amatch.c
/*
 * amatch.c - Gawk extension to access the TRE approximate regex.
 * Copyright (C) 2018 Cam Webb, <cw@camwebb.info>
 * Distributed under the GNU Pulbic Licence v3
 */

// Minimal headers:
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <gawkapi.h>
#include <tre/tre.h>

#define NSUBMATCH 10

// Gawkextlib boilerplate:
static const gawk_api_t *api;	/* for convenience macros to work */
static awk_ext_id_t ext_id;
int plugin_is_GPL_compatible;

// Main amatch() function definition
static awk_value_t * do_amatch(int nargs, awk_value_t *result, \
                            struct awk_ext_func *unused)
{
  // Variables for reading awk function's arguments
  awk_value_t re;
  awk_value_t str;
  awk_value_t costs;
  awk_value_t substr;
  
  // Read 3rd argument
  if (!get_argument(2, AWK_ARRAY, &costs)) {
    fatal(ext_id, "amatch: 3rd argument must be present, and an array");
  }

  // Read parameters for tre_regaexec()
  awk_value_t val;
  awk_value_t index;
  int i;
  char c[30];
  const char *parami[8];
  int paramv[8];
  parami[0] = "cost_ins";   paramv[0] = 1;
  parami[1] = "cost_del";   paramv[1] = 1;
  parami[2] = "cost_subst"; paramv[2] = 1;
  parami[3] = "max_cost";   paramv[3] = 3;
  parami[4] = "max_del";    paramv[4] = 3;
  parami[5] = "max_ins";    paramv[5] = 3;
  parami[6] = "max_subst";  paramv[6] = 3;
  parami[7] = "max_err";    paramv[7] = 3;
  for (i = 0; i < 8; i++) {
    make_const_string(parami[i], strlen(parami[i]), &index);
    if (get_array_element(costs.array_cookie, &index, AWK_STRING, &val)) {
      paramv[i] = atoi(val.str_value.str);
      strcpy(c,"") ;
      sprintf(c, "%s = %d", parami[i], atoi(val.str_value.str));
      warning(ext_id, c);
    }
  }

  // If the string arguments (1st and 2nd) are read. Q: die, or return NULL?
  if (!get_argument(0, AWK_STRING, &re))
    fatal(ext_id, "amatch: first parameter not found; must be a string");
  if (!get_argument(1, AWK_STRING, &str))
    fatal(ext_id, "amatch: second parameter not found; must be a string");
  
  // Compile regex
  regex_t preg;
  tre_regcomp(&preg, re.str_value.str, REG_EXTENDED);
    
  // Set approx amatch params
  regaparams_t params = { 0 };
  params.cost_ins   = paramv[0]; 
  params.cost_del   = paramv[1];
  params.cost_subst = paramv[2];
  params.max_cost   = paramv[3];
  params.max_del    = paramv[4];
  params.max_ins    = paramv[5];
  params.max_subst  = paramv[6];
  params.max_err    = paramv[7];

  // Create structure for details of match
  regamatch_t match ;
  match.nmatch = NSUBMATCH; // No partial match arrays needed
  match.pmatch = (regmatch_t *) malloc(NSUBMATCH * sizeof(regmatch_t));
  // match.pmatch; //   - ditto -
  //regmatch_t match.pmatch[4];
  
  // Return values
  int treret = 0;
  int rval = 0;

  // Do the approx regexp
  treret = tre_regaexec(&preg, str.str_value.str, &match, params, 0);
  char c2[20];
  sprintf(c2, "offset: %d", match.pmatch[2].rm_so); 
  warning(ext_id, c2);
  
  // Set the do_amatch() return value depending on tre_regaexec() return:
  // Return cost (Levenshtein distance) if success, -1 if no match,
  if (treret == REG_NOMATCH) rval = -1;
  else rval = match.cost;
  // Catch a "mem. not. allocated" return from tre_regaexec()
  if (treret == REG_ESPACE) {
    warning(ext_id,                                                     \
            "amatch: TRE err., mem. insufficient to complete the match.");
    return make_null_string(result);
  }

  
  // Hand the substrings over to the substring array
  // Read the SUBSEP symbol
  awk_value_t subsep;
  if (!sym_lookup("SUBSEP", AWK_STRING, &subsep))
    warning(ext_id, "amatch: Could not get SUBSEP from gawk.");

  // Read 4th argument
  if (!get_argument(3, AWK_ARRAY, &substr)) {
    warning(ext_id, "amatch: 4th argument not present or not an array");
  }
  else clear_array(substr.array_cookie);
  
  char outindex1[20];
  char outval1[20];
  awk_value_t outindex;
  awk_value_t outval;
  for (i = 1 ; i < match.nmatch; i++) {
    if (match.pmatch[i].rm_so > 0) {
      sprintf(outindex1, "%d", i);
      sprintf(outval1, "%d %.*s", match.pmatch[i].rm_so+1, match.pmatch[i].rm_eo - match.pmatch[i].rm_so, str.str_value.str + match.pmatch[i].rm_so);
      set_array_element(substr.array_cookie,                            \
                        make_const_string(outindex1, strlen(outindex1), &outindex), \
                        make_const_string(outval1, strlen(outval1), &outval));
    }
  }
  return make_number(rval, result);
}


// Gawkextlib boilerplate:

static awk_ext_func_t func_table[] = \
    {
     { "amatch", do_amatch, 3, 2, awk_false, NULL  },
    };

static awk_bool_t (*init_func)(void) = NULL;

static const char *ext_version = "0.1";

dl_load_func(func_table, amatch, "")



sf_build/make_sf_version.sh
# Script to clone the forked SourceForge gawkextlib codebase, apply
# changes from the github version (by copying files), build and test,
# and prepare for merge request.

# 1. Create the codebase

# git clone ssh://ctenolophon@git.code.sf.net/u/ctenolophon/gawkextlib \
#    u-ctenolophon-gawkextlib

cd u-ctenolophon-gawkextlib

# 2. Make the new default new extension

rm -rf aregex
./make_extension_directory.sh -I aregex "Cam Webb" "cw@camwebb.info"
cd aregex

# **Modify files**

# Automake/autoconf, etc
sed -i 's/-lgawkextlib/-lgawkextlib -ltre/g' Makefile.am
sed -i 's/AC_GAWK_EXTENSION/AC_GAWK_EXTENSION\n\nAC_CHECK_LIB([tre], [tre_regaexec], [],[echo "The TRE regex library is required"; exit -1])/g' configure.ac
sed -i 's/BuildRequires:    gcc/BuildRequires:    gcc\nBuildRequires:    tre-devel/g' packaging/gawk-aregex.spec.in

# Main code
cp -f ../../../aregex.c .
patch -i ../../aregex.c.patch aregex.c

# Man page
cp -f ../../../doc/aregex.3am doc

# Web page
cp -f ../../webTOC .
git add webTOC

# Test files
cp -f ../../../test/aregex.awk test
sed -i 's|./aregex.so|../.libs/aregex.so|g' test/aregex.awk
# note: the gawkextlib test suite uses gawk with --characters-as-bytes
cp -f ../../../test/aregex_b.ok test/aregex.ok

# ChangeLog
cp -f ../../../ChangeLog .

# README
cp -f ../../SF_README.md README

## **Test**

automake
autoconf
make
make check

## ** Commit **

# git commit -m "... message ..."
# git push

## ** Test in clean repo **

# D=`pwd`
# cd /tmp/
# git clone git://git.code.sf.net/u/ctenolophon/gawkextlib u-ctenolophon-gawkextlib
# cd u-ctenolophon-gawkextlib/aregex/
# autoreconf -i
# ./configure 
# make
# make check
# cd ../..
# rm -rf u-ctenolophon-gawkextlib/
# cd $D

## ** Request merge **

## ** Clean up **
# rm -rf u-ctenolophon-gawkextlib


sf_build/update_sf_version.sh
# Script to update SourceForge gawkextlib codebase, apply
# changes from the github version (by copying files), build and test,
# and prepare for merge request.

# Define new version here:

VERSION=1.0.1

# Get new version of original repo
rm -rf u-ctenolophon-gawkextlib
git clone ssh://ctenolophon@git.code.sf.net/u/ctenolophon/gawkextlib \
 u-ctenolophon-gawkextlib

cd u-ctenolophon-gawkextlib/

git remote add upstream https://git.code.sf.net/p/gawkextlib/code
git fetch upstream
git checkout master
git merge -m "auto" upstream/master

cd aregex/

# Change build files:
sed -E -i "s/\ [0-9]+\.[0-9]+\.[0-9]+/ $VERSION/g" configure.ac

# Change code files
cp -f ../../../aregex.c .
patch -i ../../aregex.c.patch aregex.c

# Man page
cp -f ../../../doc/aregex.3am doc

# Web page
cp -f ../../webTOC .

# Test files
cp -f ../../../test/aregex.awk test
sed -i 's|./aregex.so|../.libs/aregex.so|g' test/aregex.awk
# note: the gawkextlib test suite uses gawk with --characters-as-bytes
cp -f ../../../test/aregex_b.ok test/aregex.ok

# ChangeLog
cp -f ../../../ChangeLog .

# NEWS
cp -f ../../SF_NEWS NEWS

# README
cp -f ../../SF_README.md README

## Build

autoreconf -i
./configure 
make
make check

## ** Commit **

# git commit -m "... message ..."
# git push

## ** Test in clean repo **

# D=`pwd`
# cd /tmp/
# git clone git://git.code.sf.net/u/ctenolophon/gawkextlib u-ctenolophon-gawkextlib
# cd u-ctenolophon-gawkextlib/aregex/
# autoreconf -i
# ./configure 
# make
# make check
# cd ../..
# rm -rf u-ctenolophon-gawkextlib/
# cd $D

## ** Request merge **

## ** Clean up **
# rm -rf u-ctenolophon-gawkextlib


aregex.c
/*
 * aregex.c - Gawk extension to access the TRE approximate regex.
 * Copyright (C) 2018-9 Cam Webb, <cw@camwebb.info>
 * Distributed under the GNU Pulbic Licence v3
 */

// Minimal headers:
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <gawkapi.h>
#include <tre/tre.h>

#define MAXNSUBMATCH 20 // Max Number of parenthetical substring matches
#define DEFMAXCOST 5    // Default max_cost for match 
#define DEBUG 0         // Print debug info

// Gawkextlib boilerplate:
static const gawk_api_t *api;
static awk_ext_id_t ext_id;
int plugin_is_GPL_compatible;


// Main amatch() function definition
static awk_value_t * do_amatch(int nargs, awk_value_t *result \
                               , struct awk_ext_func *unused)
{
  int i; 
  
  // 1. Set default costs
  const char *parami[8];
  int paramv[8];
  parami[0] = "cost_ins";   paramv[0] = 1;
  parami[1] = "cost_del";   paramv[1] = 1;
  parami[2] = "cost_subst"; paramv[2] = 1;
  parami[3] = "max_cost";   paramv[3] = DEFMAXCOST;
  parami[4] = "max_del";    paramv[4] = DEFMAXCOST;
  parami[5] = "max_ins";    paramv[5] = DEFMAXCOST;
  parami[6] = "max_subst";  paramv[6] = DEFMAXCOST;
  parami[7] = "max_err";    paramv[7] = DEFMAXCOST;
  
  // 2. Read 3rd, 'costs' argument, if present
  //   (these variable declarations outside, because needed during output: )
  awk_value_t costs;
  awk_value_t simplecost;
  awk_value_t costindex;
  awk_value_t costval;
  awk_bool_t hascostarr = 0;
  
  if (nargs > 2) {
    // if just a simple integer for 3rd argument:
    if (get_argument(2, AWK_NUMBER, &simplecost)) {
      paramv[3] = (int) simplecost.num_value;
      paramv[4] = (int) simplecost.num_value;
      paramv[5] = (int) simplecost.num_value;
      paramv[6] = (int) simplecost.num_value;
      paramv[7] = (int) simplecost.num_value;
    }
    else if (get_argument(2, AWK_ARRAY, &costs)) {
      hascostarr = 1;
    
      char c[30];
      for (i = 0; i < 8; i++) {
        // create an index for reading array
        make_const_string(parami[i], strlen(parami[i]), &costindex);
        // if there is an array element with that index
        if (get_array_element(costs.array_cookie, &costindex,   \
                              AWK_STRING, &costval)) {
          // update the cost value
          paramv[i] = atoi(costval.str_value.str);
          if (DEBUG) {
            strcpy(c,"") ;
            sprintf(c, "cost %s = %d", parami[i], atoi(costval.str_value.str));
            warning(ext_id, c);
          }
        }
      }
    }
    else fatal(ext_id, "amatch: 3rd argument present, but could not be read.");
  }

  // 3. Read the string and regex arguments (1st and 2nd)
  awk_value_t re;
  awk_value_t str;
  if (!get_argument(0, AWK_STRING, &str))
    fatal(ext_id, "amatch: 1st param., the string, not found");
  if (!get_argument(1, AWK_STRING, &re))
    fatal(ext_id, "amatch: 2nd param., the regex, not found");

  // ( for wchar_t:
  //   wchar_t rew[] = L"";
  //   swprintf(rew, strlen(re.str_value.str), L"%ls", re.str_value.str); )
  
  // 4. Compile regex
  regex_t preg;
  tre_regcomp(&preg, re.str_value.str, REG_EXTENDED);

  // ( for wchar_t:
  //   tre_regwcomp(&preg, rew, REG_EXTENDED); )

  // 5. Do the match
  // set approx match params
  regaparams_t params = { 0 };
  params.cost_ins   = paramv[0]; 
  params.cost_del   = paramv[1];
  params.cost_subst = paramv[2];
  params.max_cost   = paramv[3];
  params.max_del    = paramv[4];
  params.max_ins    = paramv[5];
  params.max_subst  = paramv[6];
  params.max_err    = paramv[7];

  // create necessary structure for details of match
  regamatch_t match ;
  match.nmatch = MAXNSUBMATCH; 
  match.pmatch = (regmatch_t *) malloc(MAXNSUBMATCH * sizeof(regmatch_t));
  
  // do the approx regexp itself!
  int treret;
  treret = tre_regaexec(&preg, str.str_value.str, &match, params, 0);

  // ( for wchar_t:
  //   treret = tre_regawexec(&pregw, rew, &match, params, 0); )
  
  // set the amatch() return value depending on tre_regaexec() return
  //   1 if success, 0 if no match
  int rval = 1;
  if (treret == REG_NOMATCH) rval = 0;

  // Catch a "mem. not. allocated" return from tre_regaexec()
  if (treret == REG_ESPACE) {
    warning(ext_id,                                                     \
            "amatch: TRE err., mem. insufficient to complete the match.");
    free(match.pmatch);
    tre_regfree(&preg);
    return make_null_string(result);
  }

  // 6. If there is a cost array, set some return values (if a match)
  if ((hascostarr) && (rval)) {
    char matchcost[20]; // Single integers, max width ~= 10
    // cost
    del_array_element(costs.array_cookie,                           \
          make_const_string("cost", strlen("cost"), &costindex));
    sprintf(matchcost, "%d", match.cost);
    set_array_element(costs.array_cookie, \
          make_const_string("cost", strlen("cost"), &costindex), \
          make_const_string(matchcost, strlen(matchcost), &costval));
    // num_ins
    del_array_element(costs.array_cookie,                               \
          make_const_string("num_ins", strlen("num_ins"), &costindex));
    sprintf(matchcost, "%d", match.num_ins);
    set_array_element(costs.array_cookie, \
          make_const_string("num_ins", strlen("num_ins"), &costindex), \
          make_const_string(matchcost, strlen(matchcost), &costval));
    // num_del
    del_array_element(costs.array_cookie,                               \
          make_const_string("num_del", strlen("num_del"), &costindex));
    sprintf(matchcost, "%d", match.num_del);
    set_array_element(costs.array_cookie, \
          make_const_string("num_del", strlen("num_del"), &costindex), \
          make_const_string(matchcost, strlen(matchcost), &costval));
    // num_subst
    del_array_element(costs.array_cookie,                               \
          make_const_string("num_subst", strlen("num_subst"), &costindex));
    sprintf(matchcost, "%d", match.num_subst);
    set_array_element(costs.array_cookie, \
          make_const_string("num_subst", strlen("num_subst"), &costindex), \
          make_const_string(matchcost, strlen(matchcost), &costval));
  }
  
  // 7. Set 4th argument array, for matched substrings, if present
  //    and if a match found
  if ((nargs == 4) && (rval)) {
    awk_value_t substr; 
    // read 4th argument
    if (!get_argument(3, AWK_ARRAY, &substr)) {
      warning(ext_id, "amatch: Could not read 4th argument.");
    }
    else clear_array(substr.array_cookie);

    // hand the TRE substrings over to the gawk substring array
    char outindexc[20];
    int maxsubsize = 0;
    // first, find max size of substring
    for (i = 0; i < MAXNSUBMATCH ; i++)
      if (match.pmatch[i].rm_so != -1)
        if (maxsubsize < match.pmatch[i].rm_eo - match.pmatch[i].rm_so)
          maxsubsize = match.pmatch[i].rm_eo - match.pmatch[i].rm_so;
    // dimension the substring:
    char outvalc[maxsubsize+1];
    awk_value_t outindexp;
    awk_value_t outvalp;

    for (i = 0 ; i < (int) match.nmatch; i++) {
      if (match.pmatch[i].rm_so != -1) {
        sprintf(outindexc, "%d", i);
        // ( "%d %.*s", match.pmatch[i].rm_so+1, ... gives position
        //   by bytes, not by chars )
        sprintf(outvalc, "%.*s",                                 \
                match.pmatch[i].rm_eo - match.pmatch[i].rm_so,   \
                str.str_value.str + match.pmatch[i].rm_so);
        set_array_element(substr.array_cookie,                        
          make_const_string(outindexc, strlen(outindexc), &outindexp), \
          make_const_string(outvalc, strlen(outvalc), &outvalp));
      }
    }
  }
  
  free(match.pmatch);
  tre_regfree(&preg);
  return make_number(rval, result);
}


// Gawkextlib boilerplate:

static awk_ext_func_t func_table[] = \
  {
    { "amatch", do_amatch, 4, 2, awk_false, NULL  },
  };

static awk_bool_t (*init_func)(void) = NULL;

static const char *ext_version = "0.1";

dl_load_func(func_table, amatch, "")
